File: libraries/react-native/tests/core/paywall-runtime-config.test.ts
import { SubscriptionProduct } from "../../src/core/entities/product";
import {
  buildPaywallRuntimeConfig,
  mapStoreIntervalToPeriod,
} from "../../src/core/paywalls/paywall-runtime-config";
import type { PaywallReleaseRuntime } from "../../src/core/paywalls/paywall-service";
import { describe, expect, it } from "../helpers/effect-vitest";

function makeSubscriptionProduct(overrides: {
  id: string;
  slug: string;
  displayName?: string;
  description?: string;
  displayPrice?: string;
  price?: number;
  currency?: string;
  interval?: string;
}): SubscriptionProduct {
  return new SubscriptionProduct(
    overrides.id,
    overrides.slug,
    overrides.displayName ?? "Name",
    overrides.description ?? "Description",
    overrides.displayName ?? "Display Name",
    overrides.displayPrice ?? "$9.99",
    overrides.price ?? 9.99,
    overrides.currency ?? "USD",
    "autoRenewable",
    "ios",
    overrides.interval ?? "month",
  );
}

function makeRuntime(overrides: Partial<PaywallReleaseRuntime> = {}): PaywallReleaseRuntime {
  return {
    contentHash: "abc123",
    productSlugs: ["yearly", "monthly"],
    variables: { accentColor: "#16a34a" },
    ...overrides,
  };
}

describe("mapStoreIntervalToPeriod", () => {
  it("maps ISO-8601 and keyword intervals to contract periods", () => {
    expect(mapStoreIntervalToPeriod("P1M")).toBe("month");
    expect(mapStoreIntervalToPeriod("p1y")).toBe("year");
    expect(mapStoreIntervalToPeriod("P1W")).toBe("week");
    expect(mapStoreIntervalToPeriod("P7D")).toBe("week");
    expect(mapStoreIntervalToPeriod("monthly")).toBe("month");
    expect(mapStoreIntervalToPeriod("annual")).toBe("year");
    expect(mapStoreIntervalToPeriod("lifetime")).toBe("lifetime");
  });

  it("returns undefined for unknown or missing intervals", () => {
    expect(mapStoreIntervalToPeriod(undefined)).toBeUndefined();
    expect(mapStoreIntervalToPeriod("")).toBeUndefined();
    expect(mapStoreIntervalToPeriod("P3M")).toBeUndefined();
    expect(mapStoreIntervalToPeriod("fortnightly")).toBeUndefined();
  });
});

describe("buildPaywallRuntimeConfig", () => {
  it("maps resolved products onto the contract §7.1 shape", () => {
    const config = buildPaywallRuntimeConfig({
      runtime: makeRuntime(),
      productsBySlug: {
        yearly: makeSubscriptionProduct({
          id: "store_yearly",
          slug: "yearly",
          displayName: "Yearly Pro",
          description: "Best value",
          displayPrice: "$59.99",
          price: 59.99,
          currency: "USD",
          interval: "year",
        }),
        monthly: makeSubscriptionProduct({
          id: "store_monthly",
          slug: "monthly",
          displayPrice: "$9.99",
          price: 9.99,
          interval: "month",
        }),
      },
      platform: "ios",
      locale: "en-US",
    });

    expect(config.products).toEqual([
      {
        id: "store_yearly",
        slug: "yearly",
        displayName: "Yearly Pro",
        description: "Best value",
        price: 59.99,
        priceString: "$59.99",
        currencyCode: "USD",
        period: "year",
        trialPeriod: undefined,
      },
      expect.objectContaining({
        id: "store_monthly",
        slug: "monthly",
        priceString: "$9.99",
        period: "month",
      }),
    ]);
    expect(config.variables).toEqual({ accentColor: "#16a34a" });
    expect(config.locale).toBe("en-US");
    expect(config.platform).toBe("ios");
    expect(config.defaultSelectedProductId).toBe("store_yearly");
  });

  it("skips slugs the store could not resolve and reports them", () => {
    const skipped: string[] = [];
    const config = buildPaywallRuntimeConfig({
      runtime: makeRuntime({ productSlugs: ["missing", "monthly", "unknown"] }),
      productsBySlug: {
        missing: null,
        monthly: makeSubscriptionProduct({
          id: "store_monthly",
          slug: "monthly",
        }),
      },
      platform: "android",
      locale: undefined,
      onSkippedProductSlug: (slug) => skipped.push(slug),
    });

    expect(skipped).toEqual(["missing", "unknown"]);
    expect(config.products.map((product) => product.id)).toEqual(["store_monthly"]);
    expect(config.defaultSelectedProductId).toBe("store_monthly");
  });

  it("still produces a config with an empty products list", () => {
    const config = buildPaywallRuntimeConfig({
      runtime: makeRuntime({ productSlugs: ["missing"] }),
      productsBySlug: { missing: null },
      platform: "unknown",
      locale: undefined,
    });

    expect(config.products).toEqual([]);
    expect(config.defaultSelectedProductId).toBeUndefined();
    expect(config.platform).toBeUndefined();
    expect(config.locale).toBeUndefined();
    expect(config.variables).toEqual({ accentColor: "#16a34a" });
  });

  it("omits empty description/currency and non-finite price", () => {
    const config = buildPaywallRuntimeConfig({
      runtime: makeRuntime({ productSlugs: ["monthly"] }),
      productsBySlug: {
        monthly: makeSubscriptionProduct({
          id: "store_monthly",
          slug: "monthly",
          description: "",
          currency: "",
          price: Number.NaN,
        }),
      },
      platform: "ios",
      locale: "en-US",
    });

    const [product] = config.products;
    expect(product?.description).toBeUndefined();
    expect(product?.currencyCode).toBeUndefined();
    expect(product?.price).toBeUndefined();
  });
});

File: libraries/android/sdk/src/test/java/com/voidhash/sdk/paywall/PaywallRuntimeConfigTest.kt
package com.voidhash.sdk.paywall

import com.voidhash.core.paywall.PaywallBridge
import com.voidhash.sdk.api.ResolvedPaywall
import com.voidhash.sdk.billing.VoidhashProduct
import org.json.JSONObject
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNull
import kotlin.test.assertTrue

internal fun testProduct(
    id: String = "pro_monthly",
    slug: String = "pro-monthly",
    displayPrice: String = "59,99 €",
    billingPeriod: String? = "P1M",
) = VoidhashProduct(
    id = id,
    slug = slug,
    name = "Pro Monthly",
    description = "Pro access",
    displayName = "Pro",
    displayPrice = displayPrice,
    price = 59.99,
    currency = "EUR",
    type = "subs",
    billingPeriod = billingPeriod,
    googlePlayOfferToken = "offer-monthly",
)

internal fun testResolvedPaywall(productSlugs: List<String> = listOf("pro-monthly")) =
    ResolvedPaywall(
        locationSlug = "onboarding",
        htmlUrl = "https://cdn.voidhash.com/pw.html",
        releaseId = "rel_1",
        productSlugs = productSlugs,
        variables = mapOf("headline" to "Go Pro"),
    )

class PaywallRuntimeConfigTest {
    @Test
    fun `maps ISO-8601 billing periods`() {
        assertEquals("month", mapStoreIntervalToPeriod("P1M"))
        assertEquals("year", mapStoreIntervalToPeriod("P1Y"))
        assertEquals("week", mapStoreIntervalToPeriod("P1W"))
        assertEquals("week", mapStoreIntervalToPeriod("P7D"))
    }

    @Test
    fun `maps keyword variants case-insensitively`() {
        assertEquals("month", mapStoreIntervalToPeriod(" Monthly "))
        assertEquals("year", mapStoreIntervalToPeriod("ANNUAL"))
        assertEquals("lifetime", mapStoreIntervalToPeriod("lifetime"))
    }

    @Test
    fun `unknown and missing intervals map to null`() {
        assertNull(mapStoreIntervalToPeriod(null))
        assertNull(mapStoreIntervalToPeriod(""))
        assertNull(mapStoreIntervalToPeriod("P3D"))
    }

    @Test
    fun `builds the runtime config from store products`() {
        val config = buildPaywallRuntimeConfig(
            paywall = testResolvedPaywall(),
            productsBySlug = mapOf("pro-monthly" to testProduct()),
            locale = "de-DE",
        )

        assertEquals(1, config.products.size)
        assertEquals("pro_monthly", config.defaultSelectedProductId)
        assertEquals("android", config.platform)
        assertEquals("de-DE", config.locale)
        assertEquals("Go Pro", config.variables["headline"])

        val product = config.products.single()
        assertEquals("pro-monthly", product.slug)
        assertEquals("59,99 €", product.priceString)
        assertEquals("EUR", product.currencyCode)
        assertEquals("month", product.period)
        assertNull(product.trialPeriod)
    }

    @Test
    fun `a trial period is omitted from the wire until the store maps one`() {
        val config = buildPaywallRuntimeConfig(
            paywall = testResolvedPaywall(),
            productsBySlug = mapOf("pro-monthly" to testProduct()),
            locale = null,
        )

        val product = JSONObject(PaywallEnvelope.configureMessage(config))
            .getJSONObject("payload")
            .getJSONArray("products")
            .getJSONObject(0)

        assertFalse(product.has("trialPeriod"))
        assertEquals(
            "P1W",
            config.products.single().copy(trialPeriod = "P1W").toJson().getString("trialPeriod"),
        )
    }

    @Test
    fun `unresolved product slugs are reported and skipped`() {
        val skipped = mutableListOf<String>()

        val config = buildPaywallRuntimeConfig(
            paywall = testResolvedPaywall(listOf("pro-monthly", "pro-annual")),
            productsBySlug = mapOf("pro-monthly" to testProduct()),
            locale = null,
            onSkippedProductSlug = skipped::add,
        )

        assertEquals(listOf("pro-annual"), skipped)
        assertEquals(1, config.products.size)
        assertNull(config.locale)
    }

    @Test
    fun `the configure envelope is ASCII-safe`() {
        val message = PaywallEnvelope.configureMessage(
            buildPaywallRuntimeConfig(
                paywall = testResolvedPaywall(),
                productsBySlug = mapOf("pro-monthly" to testProduct()),
                locale = null,
            ),
        )

        assertTrue(message.all { it.code < 0x80 })
        assertTrue(message.contains("\\u20ac"))

        val decoded = JSONObject(message)
        assertEquals("configure", decoded.getString("type"))
        assertEquals(PaywallBridge.VERSION, decoded.getInt("version"))
        assertEquals(
            "59,99 €",
            decoded.getJSONObject("payload").getJSONArray("products").getJSONObject(0)
                .getString("priceString"),
        )
    }

    @Test
    fun `the inbound delivery script round-trips localized prices`() {
        val message = PaywallEnvelope.statusMessage("failed", error = "59,99 €")
        val script = PaywallBridge.inboundScript(message)

        assertTrue(script.startsWith("(function(){const d=atob('"))
        assertTrue(script.all { it.code < 0x80 })
        assertFalse(script.contains("€"))
    }
}
